lib/LLVMIRCodeGen/LLVMIRGen.cpp

```diff
@@ -993,11 +993,17 @@ void LLVMIRGen::emitDataParallelKernel(
 /// \param buf the buffer operand to be checked for overlaps with the \p bundle.
 static bool isOverlappingWithAnyBundleBufferOperands(
     AllocationsInfo &allocationsInfo,
-    llvm::SmallVectorImpl<const Instruction *> &bundle, Value *buf) {
+    llvm::SmallVectorImpl<const Instruction *> &bundle,
+    const Instruction::Operand &op) {
+  auto *buf = op.first;
   auto addr1 = allocationsInfo.allocatedAddress_[buf];
   auto size1 = buf->getSizeInBytes();
   for (auto bi : bundle) {
     for (auto bop : bi->getOperands()) {
+      // No new operand aliases with an input operands used before.
```

Comment:
IIUC, in order to honor "noalias" property, the following requirements must be met:
1) Output operands of a new instruction do not alias with any input operands of bundle instructions
2) Input operands of a new instruction do not alias with any output operand of bundle instructions
This, I believe the code should be like this:
```
  for (auto bi : bundle) {
    for (auto bop : bi->getOperands()) {
      // No new operand output aliases with input operands used before.
      // No new operand input aliases with output operands defined before.
      if (bop.second == OperandKind::In && op.second == OperandKind::In) {
        continue;
      }
      auto buf2 = bop.first;
      auto addr2 = allocationsInfo.allocatedAddress_[buf2];
      auto size2 = buf2->getSizeInBytes();
      if ((addr1 >= addr2 && addr1 < addr2 + size2) ||
          (addr2 >= addr1 && addr2 < addr1 + size1)) {
        return true;
      }
    }
  }
```
Or do you think it's too restrictive?
This code passed unit tests locally and solved https://github.com/pytorch/glow/issues/5534 problem for me.

Reply:
Actually, the code I just proposed has already been implemented here: https://github.com/pytorch/glow/pull/5370

Reply:
@aturetsk Yeah, I agree that this is a safe thing to do. I pushed the new version. It ensures correctness.

But I suspect it may also affect performance as it is more pessimistic and makes bundles smaller. As a result, generated stacked kernels have noalias operands, but they combine less operations into one kernel. Thus, more stacked kernels will be generated and invoked and more passes over tensors will be done by them, potentially trashing the cache. 

But I think we should first do this fix to ensure correctness and worry about the performance later if it really becomes an issue.

Reply:
@aturesk @andreas-hommel OK to land this for now?

Reply:
Ok by me.

Reply:
Agreed, correctness always comes first